### ISSUE TITLE
Registration evaluation

### ISSUE THREAD

**Issue by dyollb** [uncbiag/uniGradICON#38]
Thanks for making this tool available. The slicer integration is great!

I have been evaluating inter-subject whole-head registration (T1 MRI to T1 MRI).

overall I am impressed, but trying to extend my experiments I have the following observations. Not sure how to overcome these obstacles
- the deformation fields can be very extreme (non-smooth), unfortunately this cannot be controlled
- I frequently get problems (high frequency spiky deformation) at the skin (I don't mask the image), e.g. see the alien eyes (warped segmentation)
- I see much better performance on T1 images from Philips scanners (in my datasets) than for GE/Siemens. Was it trained more on Philips images (e.g. IXI)?
- when I have bright areas (electrodes) in the image, this can destroy the result (masking those areas helps though), see images below

| Extreme Deformation at Skin/Boundary| Electrodes | Failed Due to Electrodes |
|----------|----------|----------|
| ![Image](https://github.com/user-attachments/assets/7cbad513-cbb5-4dd9-a850-3fb1a5a0e30a) | ![Image](https://github.com/user-attachments/assets/80a9f7b2-d1f9-4042-a1b1-dd3a7a76c1bb) | ![Image](https://github.com/user-attachments/assets/10a22f99-1c1c-4cf6-ad76-e18f795551c3) |

**Comment by dyollb** [comment]
Would it help to reduce the extreme local deformation if i
- pre-align the images using sn affine registration 
- mask the image (foreground)?

**Comment by lintian-a** [comment]
Hi @dyollb ,

Thanks for your interest in our work! 

> 1. the deformation fields can be very extreme (non-smooth), unfortunately this cannot be controlled

 I might need more context to determine the right solution. Could you clarify which extreme local deformation you are referring to in the result? Is it in area 1 or ares 2 in the following image.

 ![Image](https://github.com/user-attachments/assets/b64f470b-e2bb-4c6f-ba31-5abcf1f8e435)

> 2. I frequently get problems (high frequency spiky deformation) at the skin (I don't mask the image), e.g. see the alien eyes (warped segmentation)

Are you referring to the deformation in area 2 in the above image? Could you also post the screenshot of the source and target images? 

> 3. I see much better performance on T1 images from Philips scanners (in my datasets) than for GE/Siemens. Was it trained more on Philips images (e.g. IXI)?

uniGradICON is trained with HCP, COPDGene, OAI and L2R Abdomen datasets together. We did not use IXI for training. Running systematic analysis on how uniGradICON performs across device would be very interesting! Unfortunately, I haven't run such an experiment thus I am not sure what the reason is for your observation.
 
> 4. when I have bright areas (electrodes) in the image, this can destroy the result (masking those areas helps though), see images below

It looks like an orientation issue. Currently, uniGradICON assumes the target image and source image have the same orientation. The orientation could be any one but it has to be the same for both images. We have implemented a reorient function in #13  but haven't fully tested it yet. Could you help confirm whether the source and target images have different orientations for the case in the third image you were showing?

**Comment by dyollb** [comment]
Sorry for not replying sooner. I was away for some days.

> Hi [@dyollb](https://github.com/dyollb) ,
> 
> Thanks for your interest in our work!
> 
> > 1. the deformation fields can be very extreme (non-smooth), unfortunately this cannot be controlled
> 
> I might need more context to determine the right solution. Could you clarify which extreme local deformation you are referring to in the result? Is it in area 1 or ares 2 in the following image.

Yes, exactly. For example, I get this unrealistic large deformation around the eyes in many cases, as seen in 1.

> 
> ![Image](https://github.com/user-attachments/assets/b64f470b-e2bb-4c6f-ba31-5abcf1f8e435)
> 
> > 2. I frequently get problems (high frequency spiky deformation) at the skin (I don't mask the image), e.g. see the alien eyes (warped segmentation)
> 
> Are you referring to the deformation in area 2 in the above image? Could you also post the screenshot of the source and target images?

Yes, exactly.

> 
> > 3. I see much better performance on T1 images from Philips scanners (in my datasets) than for GE/Siemens. Was it trained more on Philips images (e.g. IXI)?
> 
> uniGradICON is trained with HCP, COPDGene, OAI and L2R Abdomen datasets together. We did not use IXI for training. Running systematic analysis on how uniGradICON performs across device would be very interesting! Unfortunately, I haven't run such an experiment thus I am not sure what the reason is for your observation. I will reply in this post if I have a chance to run such an analysis with suitable datasets in the future.

I am using the [Calgary-Campinas (CC259)](https://sites.google.com/view/calgary-campinas-dataset/home) dataset, which includes 1.5T and 3T T1 images from GE, Siemens and Philips scanners.
It seems the HCP dataset was acquired using Siemens 3T scanners. It might also be due to a slightly different field-of-view: in the CC259 dataset the Philipps images have more "empty space" around the head than GE/Siemens images.

> 
> > 4. when I have bright areas (electrodes) in the image, this can destroy the result (masking those areas helps though), see images below
> 
> It looks like an orientation issue. Currently, uniGradICON assumes the target image and source image have the same orientation. The orientation could be any one but it has to be the same for both images. We have implemented a reorient function in [#13](https://github.com/uncbiag/uniGradICON/pull/13) but haven't fully tested it yet. Thus, it is not in the current version. Could you help confirm whether the source and target images have different orientations for the case in the third image you were showing?

I first thought it was an orientation issue. But all my images are in RAS orientation and looking in the same direction (see images below). However, after manually removing the electrodes from the MRI the result was immediately much better: I segmented them and then painted that region to similar values as the background/air.

| Moving | Fixed |
|----------|----------|
| ![Image](https://github.com/user-attachments/assets/cae1adb7-9e29-4866-8b52-5a833f3cea98)   | ![Image](https://github.com/user-attachments/assets/1ecf0f31-6ce2-47ad-973f-f37a6de9d337)  |

As a follow-up, I will try to mask the background in the moving and fixed images before registration (which should prevent finding spurious matches at the boundary leading to large deformation, e.g., in the eyes).

**Comment by lintian-a** [comment]
I'm glad that the segmentation preprocessing works for you! Our method currently will try to match everything in the input image pairs. Thus, it could produce unwanted deformation if the two images contain different contents. 

If you are using the command line interface, you can pass the segmentation mask (or ROI mask) as an input to specify the area you want the network to register. Sorry that we haven't exposed this parameter yet for the current version of the slicer extension. 

In addition, I don't know whether this is related to the issue you were having. We have an intensity preprocessing step before the input image pairs are passed to the neural network. For MRI, we use Min-Max normalization to normalize the intensity to [0,1], and we use the 99% percentile of intensity as the max value. Is it possible that the electrodes' intensity could provide an unexpected appearance after the intensity preprocessing?

**Comment by dyollb** [comment]
The segmentation masks seem to help constrain the boundary deformation. Overall this also improves the alignment further inside. I guess if the models were trained  with unmasked images, e.g. full head mri instead of skullstripped brain images.

**Comment by lintian-a** [comment]
Thanks for the update! I might know where the issue is. I see some structures in the fixed image that are not that obvious in the original contrast, but they probably will stand out after the intensity processing. 

<img width="534" alt="Image" src="https://github.com/user-attachments/assets/0a428793-ced5-4e54-b5f9-3626a2f77776" />

Could you help confirm that by running this [colab notebook](https://colab.research.google.com/drive/1JuFL113WN3FHCoXG-4fiBTWIyYpwGyGy?usp=sharing)? It exposes the processing and the images we pass to the network after the processing. You can try it with your data and plot the image pairs after the processing.

uniGradICON will try to align everything between the two images. Thus, if there are different contents (caused by different field of views or acquisitions) contained in the image pair, it will also try to stretch the boundary, causing undesired deformations.

**Comment by dyollb** [comment]
Thanks. This is the before intensity processing (left: moving, right: fixed):

![Image](https://github.com/user-attachments/assets/847a616b-da6f-4d49-b07f-a609492b982c)

This is after pre-processing

![Image](https://github.com/user-attachments/assets/88fc0ad3-b35f-4211-8db6-bba6c9873fec)

**Comment by dyollb** [comment]
so I have a reasonable workaround. thanks for all your support!

but out of curiosity before closing this issue, do you think it would be worth including unmasked datasets in the training data?
or would you recommend I just train a UNet to predict the foreground mask (head versus background) better than e.g. "Otsu thresholding + morphological operations"?

**Comment by lintian-a** [comment]
I changed the contrast of the post-processing image and here is how it looks like

<img width="1829" alt="Image" src="https://github.com/user-attachments/assets/8e25baa2-e9bb-4fda-92ff-7536e62d7a60" />

There is some noise in the background for the target image. If you are fine-tuning a model based on uniGradICON for your dataset, I suggest using the masked image, which cleans the background. 

> Would you recommend I just train a UNet to predict the foreground mask (head versus background) better than e.g. "Otsu thresholding + morphological operations"

I think it depends on your needs. If you already have a robust traditional segmentation method, you can use it as a pre-processing before uniGradICON. If the traditional segmentation is very slow and the run-time is critical for your application, you might want to train a segmentation network.

**Comment by dyollb** [comment]
👍
